### PR TITLE
feat: Entity explorer revamp

### LINKED ISSUE #9671: [Task] - Entity explorer - Pages section

### LINKED ISSUE #9672: [Task] - Entity explorer - Datasource section

### LINKED ISSUE #9688: [Task] - Add file operations to omnibar

### LINKED ISSUE #9689: [Task] - Entity explorer - Files section

### PR DESCRIPTION
**Pull request by albinAppsmith**
## Description
A complete refresh to our entity explorer. 

### Problems with the existing explorer
- It shows the entities of all the pages in the application making it difficult to use when an application grows.
- It re-renders on every(almost) action done on the UI.
- It fails to show the queries created under a datasource.
- Generic icons makes it difficult to differentiate between kind of queries.
- Widgets take up a lot of real estate and it is expanded on every widget selection, pushing the relevant items further down.

### Changes
- We no longer show the entities from all the pages but just of the page, user is currently on.
- Performance improvements by writing better selectors and memoizing components.
- Queries now have the same icon as the datasource except that they are boxed.
- We have a new section which shows all the queries and JS Objects. They are grouped and sorted by default.
- Widgets section is expanded only on demand.
- Action creation will now go via the omnibar.
- Bindings are no longer under the entity but is shown in a separate popover.

Fixes #9689 
Fixes #9671 
Fixes #9672 
Fixes #9688

## Type of change

- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

## How Has This Been Tested?

- Manual
- Cypress

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: feat/organize_entity_exp 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 55.01 **(-0.02)** | 36.53 **(-0.04)** | 34.74 **(-0.2)** | 55.41 **(-0.03)**
 :green_circle: | app/client/src/components/ads/TreeDropdown.tsx | 59.26 **(0)** | 61.11 **(2.29)** | 36.84 **(0)** | 57.69 **(0)**
 :green_circle: | app/client/src/components/editorComponents/Sidebar.tsx | 83.15 **(1.54)** | 44.12 **(0)** | 62.5 **(6.25)** | 83.33 **(1.62)**
 :green_circle: | app/client/src/components/editorComponents/ActionCreator/index.tsx | 39.11 **(0)** | 7.08 **(0.12)** | 40 **(0)** | 38.98 **(0)**
 :red_circle: | app/client/src/components/editorComponents/CodeEditor/generateQuickCommands.tsx | 26.25 **(-3.08)** | 7.14 **(0.47)** | 6.25 **(0)** | 25.32 **(-3.06)**
 :red_circle: | app/client/src/components/editorComponents/GlobalSearch/Description.tsx | 21.05 **(-0.28)** | 29.63 **(0)** | 0 **(0)** | 22.22 **(-0.32)**
 :sparkles: :new: | **app/client/src/components/editorComponents/GlobalSearch/GlobalSearchHooks.tsx** | **71** | **21.88** | **50** | **66.28**
 :red_circle: | app/client/src/components/editorComponents/GlobalSearch/HelpBar.tsx | 93.55 **(0.69)** | 66.67 **(-8.33)** | 90.91 **(0.91)** | 92.31 **(0.31)**
 :red_circle: | app/client/src/components/editorComponents/GlobalSearch/SearchBox.tsx | 25 **(-0.37)** | 18.75 **(-0.6)** | 0 **(0)** | 28.81 **(-0.5)**
 :red_circle: | app/client/src/components/editorComponents/GlobalSearch/SearchResults.tsx | 23.4 **(-1.41)** | 26.97 **(-2.66)** | 0 **(0)** | 24.26 **(-1.12)**
 :green_circle: | app/client/src/components/editorComponents/GlobalSearch/index.tsx | 52.52 **(0.23)** | 17.19 **(2.57)** | 34.48 **(-2.83)** | 54.98 **(1.16)**
 :red_circle: | app/client/src/components/editorComponents/GlobalSearch/utils.tsx | 67.33 **(-2.27)** | 23.58 **(-1.2)** | 53.57 **(-7.3)** | 66.94 **(-2.06)**
 :red_circle: | app/client/src/constants/messages.ts | 74.85 **(-0.09)** | 100 **(0)** | 25.45 **(-0.3)** | 80.29 **(0.07)**
 :red_circle: | app/client/src/constants/routes.ts | 64.96 **(-2.56)** | 24.36 **(-2.56)** | 22.22 **(-7.41)** | 61.27 **(-2.95)**
 :red_circle: | app/client/src/icons/MenuIcons.tsx | 58.46 **(-4.25)** | 100 **(0)** | 6.9 **(-8.48)** | 58.46 **(-4.25)**
 :green_circle: | app/client/src/pages/Editor/EditorHeader.tsx | 86.52 **(0.29)** | 53.01 **(2.33)** | 62.5 **(0)** | 86.03 **(0.32)**
 :red_circle: | app/client/src/pages/Editor/GlobalHotKeys.tsx | 61.42 **(-0.48)** | 28.57 **(0)** | 50 **(-1.11)** | 60.66 **(-0.5)**
 :green_circle: | app/client/src/pages/Editor/MainContainer.tsx | 93.1 **(4.73)** | 100 **(16.67)** | 66.67 **(11.11)** | 92.59 **(5.75)**
 :sparkles: :new: | **app/client/src/pages/Editor/Explorer/Datasources.tsx** | **92.11** | **62.5** | **50** | **91.43**
 :green_circle: | app/client/src/pages/Editor/Explorer/EntityExplorer.tsx | 88.89 **(9.79)** | 66.67 **(29.17)** | 50 **(16.67)** | 88.46 **(8.77)**
 :red_circle: | app/client/src/pages/Editor/Explorer/ExplorerIcons.tsx | 75 **(-14.06)** | 30 **(-20)** | 6.67 **(-13.33)** | 78.82 **(-9.89)**
 :red_circle: | app/client/src/pages/Editor/Explorer/JSDependencies.tsx | 81.82 **(-3.03)** | 100 **(0)** | 42.86 **(-7.14)** | 85.71 **(-3.58)**
 :green_circle: | app/client/src/pages/Editor/Explorer/helpers.tsx | 76.47 **(12.83)** | 50 **(15.38)** | 50 **(25)** | 73.33 **(14.71)**
 :red_circle: | app/client/src/pages/Editor/Explorer/hooks.ts | 22.87 **(-14.36)** | 19.83 **(11.38)** | 12.24 **(-41.25)** | 22.78 **(-17.36)**
 :red_circle: | app/client/src/pages/Editor/Explorer/Actions/ActionEntity.tsx | 48.39 **(-16.61)** | 0 **(0)** | 0 **(0)** | 51.72 **(-13.28)**
 :green_circle: | app/client/src/pages/Editor/Explorer/Actions/ActionEntityContextMenu.tsx | 40.38 **(1.25)** | 0 **(0)** | 0 **(0)** | 40.38 **(1.25)**
 :green_circle: | app/client/src/pages/Editor/Explorer/Actions/MoreActionsMenu.tsx | 34.78 **(1.45)** | 25 **(0)** | 0 **(0)** | 37.21 **(1.5)**
 :red_circle: | app/client/src/pages/Editor/Explorer/Actions/helpers.tsx | 36.59 **(-10.29)** | 0 **(-34.55)** | 0 **(-19.05)** | 37.14 **(-11.95)**
 :red_circle: | app/client/src/pages/Editor/Explorer/Datasources/DatasourceEntity.tsx | 39.13 **(-6.82)** | 0 **(0)** | 0 **(0)** | 39.13 **(-6.82)**
 :green_circle: | app/client/src/pages/Editor/Explorer/Entity/Collapse.tsx | 100 **(11.11)** | 100 **(25)** | 100 **(0)** | 100 **(0)**
 :green_circle: | app/client/src/pages/Editor/Explorer/Entity/CollapseToggle.tsx | 100 **(8.33)** | 100 **(25)** | 100 **(0)** | 100 **(0)**
 :green_circle: | app/client/src/pages/Editor/Explorer/Entity/EntityProperties.tsx | 52.63 **(31.42)** | 19.23 **(19.23)** | 52.63 **(52.63)** | 50.96 **(29.42)**
 :green_circle: | app/client/src/pages/Editor/Explorer/Entity/EntityProperty.tsx | 57.14 **(-1.95)** | 75 **(41.67)** | 0 **(0)** | 60.61 **(1.52)**
 :red_circle: | app/client/src/pages/Editor/Explorer/Entity/Name.tsx | 54.26 **(-8.09)** | 32.08 **(-2.61)** | 26.32 **(-12.14)** | 53.33 **(-8.4)**
 :green_circle: | app/client/src/pages/Editor/Explorer/Entity/Placeholder.tsx | 100 **(45.45)** | 100 **(0)** | 100 **(100)** | 100 **(45.45)**
 :green_circle: | app/client/src/pages/Editor/Explorer/Entity/index.tsx | 90.59 **(0.47)** | 82.86 **(6.07)** | 84.21 **(0.88)** | 90.12 **(0.38)**
 :sparkles: :new: | **app/client/src/pages/Editor/Explorer/Files/index.tsx** | **75** | **16.67** | **50** | **80**
 :green_circle: | app/client/src/pages/Editor/Explorer/JSActions/JSActionContextMenu.tsx | 37.84 **(2.36)** | 0 **(0)** | 0 **(0)** | 38.89 **(2.22)**
 :red_circle: | app/client/src/pages/Editor/Explorer/JSActions/JSActionEntity.tsx | 60.87 **(-4.35)** | 100 **(0)** | 0 **(0)** | 60.87 **(-4.35)**
 :green_circle: | app/client/src/pages/Editor/Explorer/JSActions/MoreJSActionsMenu.tsx | 33.33 **(1.75)** | 33.33 **(0)** | 0 **(0)** | 36.11 **(1.82)**
 :sparkles: :new: | **app/client/src/pages/Editor/Explorer/Pages/index.tsx** | **80.36** | **62.5** | **38.46** | **80.36**
 :red_circle: | app/client/src/pages/Editor/Explorer/Widgets/WidgetContextMenu.tsx | 70.83 **(-1.9)** | 14.29 **(0)** | 50 **(-16.67)** | 70.21 **(-1.88)**
 :red_circle: | app/client/src/pages/Editor/Explorer/Widgets/WidgetEntity.tsx | 97.62 **(-0.3)** | 79.31 **(-3.67)** | 100 **(0)** | 100 **(0)**
 :green_circle: | app/client/src/pages/Editor/Explorer/Widgets/WidgetGroup.tsx | 100 **(6.45)** | 74 **(3.17)** | 100 **(33.33)** | 100 **(6.67)**
 :red_circle: | app/client/src/pages/Editor/SaaSEditor/constants.ts | 45.71 **(-8.58)** | 0 **(-6.67)** | 0 **(-25)** | 38.71 **(-9.68)**
 :red_circle: | app/client/src/pages/Editor/WidgetsEditor/index.tsx | 91.78 **(-1.87)** | 62.07 **(0)** | 75 **(-25)** | 91.18 **(-2.15)**
 :green_circle: | app/client/src/reducers/uiReducers/explorerReducer.ts | 48 **(4.52)** | 0 **(0)** | 10 **(10)** | 48 **(4.52)**
 :green_circle: | app/client/src/sagas/ApiPaneSagas.ts | 17.19 **(1.59)** | 1.96 **(0.14)** | 9.09 **(1.95)** | 19.39 **(1.74)**
 :green_circle: | app/client/src/sagas/DatasourcesSagas.ts | 13.51 **(0.08)** | 2.4 **(0.04)** | 8.7 **(0)** | 16.12 **(0.12)**
 :red_circle: | app/client/src/sagas/QueryPaneSagas.ts | 22.66 **(-2.56)** | 10.87 **(-0.49)** | 18.18 **(-6.82)** | 26.67 **(-2.8)**
 :red_circle: | app/client/src/selectors/entitiesSelector.ts | 46.95 **(-2.91)** | 8.53 **(-2.97)** | 26.24 **(-5.34)** | 43.61 **(-2.94)**
 :green_circle: | app/client/src/utils/helpers.tsx | 65.78 **(2.32)** | 32.22 **(2.22)** | 58.82 **(7.84)** | 63.68 **(3)**
 :x: | ~~app/client/src/pages/Editor/Explorer/Actions/ActionsGroup.tsx~~ | ~~50~~ | ~~0~~ | ~~0~~ | ~~50~~
 :x: | ~~app/client/src/pages/Editor/Explorer/Entity/CurrentPageEntityProperties.tsx~~ | ~~27.08~~ | ~~0~~ | ~~0~~ | ~~28.26~~
 :x: | ~~app/client/src/pages/Editor/Explorer/JSActions/JSActionGroup.tsx~~ | ~~63.16~~ | ~~0~~ | ~~0~~ | ~~63.16~~
 :x: | ~~app/client/src/pages/Editor/Explorer/Pages/PageEntity.tsx~~ | ~~90.91~~ | ~~40~~ | ~~60~~ | ~~90.91~~
 :x: | ~~app/client/src/pages/Editor/Explorer/Pages/PageGroup.tsx~~ | ~~82.61~~ | ~~80~~ | ~~50~~ | ~~84.44~~
 :x: | ~~app/client/src/pages/Editor/Explorer/PluginGroup/PluginGroup.tsx~~ | ~~89.66~~ | ~~65.96~~ | ~~60~~ | ~~89.29~~</details>

### CODE REVIEW

**Review by arunvjn — APPROVED** [review]
Approving for tests

**Review by arunvjn — APPROVED** [review]

**Review by arunvjn — DISMISSED** [review]

**Review by akash-codemonk — CHANGES_REQUESTED** [review]
https://user-images.githubusercontent.com/67054171/146734466-b4b565e9-7da2-4b82-a4fb-ff7fcd8b233a.mov

Could be some memoization issue.

**Comment by arunvjn** [comment]
> ds-name.mov
> 
> Could be some memoization issue.

Nice catch. Fixed this bug

**Review by arunvjn — APPROVED** [review]
Approving for tests

**Review by arunvjn — APPROVED** [review]

**Comment by SatishGandham** [comment]
@arunvjn Did you check the re-renders of the name component?

**Comment by arunvjn** [comment]
> @arunvjn Did you check the re-renders of the name component?

@SatishGandham There are multiple selectors looking for existing entity names, which are the cause of re-renders. Eg. 
_getExistingActionNames_ will re-render anytime the actions array in state changes. This is because anytime we update an aciton we get a new reference to the actions array in the state. We can work around this by maintaining a list of used entity names in the state and updating it only when a rename or new entity gets added.

**Comment by SatishGandham** [comment]
@arunvjn 
If something is rendering when it doesn't have to we have to fix it. We can prioritise which gets fixed first after running a profile. 
Entity explorer should not re-render every-time the state changes (which very frequent). This will impact the editing experience.

Regarding `existingActionNames` selector, it is only used to check if something is a valid name and not used in render. In such cases, we can directly access the store so that we don't trigger a render.

**Review by arunvjn — APPROVED** [review]
Rerun tests

**Review by arunvjn — APPROVED** [review]
Rerun tests

**Review by arunvjn — APPROVED** [review]
Rerun tests

**Review by arunvjn — APPROVED** [review]
Tests rerun

**Review by arunvjn — APPROVED** [review]
Rerun tests

File: app/client/src/pages/Editor/Explorer/JSActions/JSActionContextMenu.tsx [hunk 1]
```diff
@@ -75,6 +92,11 @@ export function JSCollectionEntityContextMenu(props: EntityContextMenuProps) {
           onSelect: editJSCollectionName,
           label: "Edit Name",
         },
+        {
+          value: "showBinding",
+          onSelect: () => showBinding(props.id, props.name),
+          label: "Show Bindings",
```
Comment (akash-codemonk) [inline app/client/src/pages/Editor/Explorer/JSActions/JSActionContextMenu.tsx:98]: Can we move this static string to constants/messages.

File: app/client/src/pages/Editor/Explorer/Widgets/WidgetEntity.tsx [hunk 2]
```diff
@@ -8,13 +8,13 @@ import { getWidgetIcon } from "../ExplorerIcons";
...
-import EntityProperties from "../Entity/EntityProperties";
+// import { ENTITY_TYPE } from "entities/DataTree/dataTreeFactory";
+// import EntityProperties from "../Entity/EntityProperties";
 import { CanvasStructure } from "reducers/uiReducers/pageCanvasStructureReducer";
-import CurrentPageEntityProperties from "../Entity/CurrentPageEntityProperties";
+// import CurrentPageEntityProperties from "../Entity/CurrentPageEntityProperties";
 import { getSelectedWidget, getSelectedWidgets } from "selectors/ui";
 import { useNavigateToWidget } from "./useNavigateToWidget";
-import { getCurrentPageId } from "selectors/editorSelectors";
+// import { getCurrentPageId } from "selectors/editorSelectors";
```
Comment (akash-codemonk) [inline app/client/src/pages/Editor/Explorer/Widgets/WidgetEntity.tsx:17]: Can we remove these commented out imports?

**Review by akash-codemonk — CHANGES_REQUESTED** [review]

**Review by akash-codemonk — APPROVED** [review]

**Pull request merged**